src/Telemetry/Microsoft.DotNet.AzureDevOpsTimelineTests/AzureDevOpsTimelineTimeoutTests.cs

```diff
@@ -0,0 +1,106 @@
+            {
+                collection.AddScoped<AzureDevOpsTimeline>();
+                return s => s.GetRequiredService<AzureDevOpsTimeline>();
+            }
+
+            [ConfigureAllParameters]
+            public static Func<IServiceProvider, InMemoryTimelineTelemetryRepository> Repository(
+                IServiceCollection collection, string project, DateTimeOffset latestTime)
+            {
+                collection.AddSingleton<ITimelineTelemetryRepository>(
+                    s => new InMemoryTimelineTelemetryRepository(
+                        new List<(string project, DateTimeOffset latestTime)> { (project, latestTime) }
+                    )
+                );
+
+                return s => (InMemoryTimelineTelemetryRepository)s.GetRequiredService<ITimelineTelemetryRepository>();
+            }
+
+            public static void Clock(IServiceCollection collection, DateTimeOffset staticClock)
+            {
+                Mock<ISystemClock> mockSystemClock = new Mock<ISystemClock>();
+                mockSystemClock.Setup(x => x.UtcNow).Returns(staticClock);
+                collection.AddSingleton(mockSystemClock.Object);
+            }
+
+            public static void Build(IServiceCollection collection, BuildAndTimeline build, HttpMessageHandler httpMessageHandler)
+            {
+                collection.AddSingleton<IAzureDevOpsClient>(client => new MockTimeoutAzureClient(new Dictionary<Build, List<Timeline>>
+                {
+                    {build.Build, build.Timelines.ToList()}
+                }, httpMessageHandler));
+                collection.AddSingleton<IBuildLogScraper, BuildLogScraper>();
+            }
+        }
+
+        [Test]
+        public async Task TestAzureDevOpsTimelineTimeout()
+        {
+            DateTimeOffset timeDatum = DateTimeOffset.Parse("2021-01-01T01:00:00Z");
+            string azdoProjectName = "public";
+            string targetBranchName = "theTargetBranch";
+            MockHttpMessageHandler httpMessageHandler = MockHttpMessageHandler.Create("some string");
+
+            BuildAndTimeline build = BuildAndTimelineBuilder.NewPullRequestBuild(1, azdoProjectName, targetBranchName)
+                .AddTimeline(TimelineBuilder.EmptyTimeline("1", timeDatum)
+                    .AddRecord("NetCore1ESPool-Internal 5", "Initialize job", "https://www.microsoft.com"))
+                .Build();
+
+            // Test setup
+            await using TestData testData = await TestData.Default
+                .WithStaticClock(timeDatum)
+                .WithBuild(build)
+                .WithHttpMessageHandler(httpMessageHandler)
+                .BuildAsync();
+
+            await testData.Controller.RunProject(azdoProjectName, 1000, CancellationToken.None);
+
+            testData.Repository.TimelineRecords.Count(record => string.IsNullOrEmpty(record.ImageName)).Should().Be(1);
+            httpMessageHandler.RequestCancelledCount.Should().Be(1);
```

Comment:
Does this rely on actual time running out?

Reply:
not anymore